src/lcd.c

```diff
@@ -0,0 +1,228 @@
+#include "lcd.h"
+#include "gpio.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+
+typedef struct  LCD_handler_s
+{
+    uint8_t rs_pin; // LOW: command.  HIGH: character.
+    uint8_t rw_pin; // LOW: write to LCD.  HIGH: read from LCD.
+    uint8_t enable_pin; // activated by a HIGH pulse.
+    uint8_t data_pins[8];
+
+    uint8_t displayfunction;
+    uint8_t displaycontrol;
+    uint8_t displaymode;
+
+    uint8_t initialized;
+
+    uint8_t numlines;
+    uint8_t row_offsets[4];
+}               LCD_handler;
+
+static LCD_handler handler;
+
+static void    pulse_enable(void)
+{
+    gpio_set(GPIOC, handler.enable_pin, 0);
+    DELAY
+    gpio_set(GPIOC, handler.enable_pin, 1);
+    DELAY
+    gpio_set(GPIOC, handler.enable_pin, 0);
+    DELAY
+}
+
+static void    write4bits(uint8_t value)
+{
+    gpio_set(GPIOB, handler.data_pins[0], (value >> 0) & 0x1);
+    gpio_set(GPIOB, handler.data_pins[1], (value >> 1) & 0x1);
+    gpio_set(GPIOB, handler.data_pins[2], (value >> 2) & 0x1);
+    gpio_set(GPIOA, handler.data_pins[3], (value >> 3) & 0x1);
+
+    pulse_enable();
+}
+
+void    lcd_clear(void)
+{
+    lcd_command(LCD_CLEARDISPLAY);
+    DELAY
```

Comment:
This looks unidiomatic. Macros that do something should look like function call: `DELAY();`. Otherwise, it's non-obvious and I have to loop up whether that's an error or just the way the macro works.